### PR TITLE
Implement each lazy map entry as a substate

### PR DESCRIPTION
**Pull request by talekhinezh**

### CODE REVIEW

**Review by iamyulong — COMMENTED** [review]
Nice!

File: radix-engine/src/engine/process.rs [hunk 1]
```diff
@@ -1198,27 +1199,25 @@ impl<'r, 'l, L: SubstateStore> Process<'r, 'l, L> {
...
                             component_address: *component_address,
                         },
                     ))
                 }
                 _ => Err(RuntimeError::LazyMapNotFound(input.mid)),
             },
-            Some((root, lazy_map)) => Ok((lazy_map, Uncommitted { root })),
+            Some((root, value)) => Ok((value, Uncommitted { root })),
         }?;
         let mut new_entry_object_refs = Self::process_entry_data(&input.value)?;
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/engine/process.rs:1216]: Ah yes, that's right, component objects should not be in the key part, as key in lazymap is mechanism by which to find component objects. But yes, this is part of larger conversation.

File: radix-engine/src/ledger/memory.rs [hunk 2]
```diff
@@ -69,20 +70,33 @@ impl SubstateStore for InMemorySubstateStore {
...
+            .insert((component_address, mid, key), value);
     }
 
-    fn get_vault(&self, component_address: &Address, vid: &Vid) -> Option<Vault> {
+    fn get_vault(&self, component_address: &Address, vid: &Vid) -> Vault {
         self.vaults
             .get(&(component_address.clone(), vid.clone()))
-            .map(Clone::clone)
+            .cloned()
+            .unwrap()
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/ledger/memory.rs:99]: good point, will add

[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/ledger/memory.rs:101]: I don't believe there is such a difference for non-existent vault id from committed/uncommitted/component_state/lazy_map. It's simply nonexistent with no other state attached. Or am I missing something?

[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/ledger/memory.rs:101]: Ah I see now, if nonexistent vault is stored in any of these.

**Review by iamyulong — APPROVED** [review]

**Pull request merged**